components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/nats/StreamingListener.java

```diff
@@ -0,0 +1,215 @@
+                            NatsConstants.DEFAULT_NATS_STREAMING_CLUSTER_ID :
+                            natsStreamingClusterId).connectionLostHandler((streamingConnection, e) -> {
+                        NatsMessageConsumer messageConsumer = NatsEndpointManager.getInstance().getMessageConsumer();
+                        try {
+                            if (connection != null) connection.close();
+                            connection = null;
+                            subscription = null;
+                            messageConsumer.consumeMessage();
+                        } catch (IOException | InterruptedException ex) {
+                            log.error("An error occurred while connecting to NATS server, consuming messages or while closing the connection. " + ex);
+                            messageConsumer.closeConnection();
+                        } catch (SynapseException ex) {
+                            log.error("Error while retrieving or injecting NATS message. " + e.getMessage(), ex);
+                        } catch (Exception ex) {
+                            log.error("Error while retrieving or injecting NATS message or closing conection. " + e.getMessage(), ex);
+                            messageConsumer.closeConnection();
+                        }
+                    });
+
+            if (Boolean.parseBoolean(natsProperties.getProperty(NatsConstants.USE_CORE_NATS_CONNECTION))) {
+                builder.natsConn(new CoreListener(subject, injectHandler, natsProperties).getNatsConnection());
+            }
+
+            if (StringUtils.isNotEmpty(connectWait)) {
+                builder.connectWait(Duration.ofSeconds(Integer.parseInt(connectWait)));
+            }
+
+            if (StringUtils.isNotEmpty(discoverPrefix)) {
+                builder.discoverPrefix(discoverPrefix);
+            }
+
+            if (StringUtils.isNotEmpty(maxPingsOut)) {
+                builder.maxPingsOut(Integer.parseInt(maxPingsOut));
+            }
+
+            if (StringUtils.isNotEmpty(pingInterval)) {
+                builder.pingInterval(Duration.ofSeconds(Integer.parseInt(pingInterval)));
+            }
+
+            if (Boolean.parseBoolean(traceConnection)) {
+                builder.traceConnection();
+            }
+
+            StreamingConnectionFactory connectionFactory = new StreamingConnectionFactory(builder.build());
+            connection = connectionFactory.createConnection();
+        }
+        return true;
+    }
+
+    /**
+     * Consume the message received and inject into the sequence.
+     *
+     * @param sequenceName the sequence to inject the message to.
+     */
+    @Override public void consumeMessage(String sequenceName)
+            throws InterruptedException, IOException, TimeoutException {
+        SubscriptionOptions.Builder subscriptionOptions = new SubscriptionOptions.Builder();
```

Comment:
What's the difference between this and the builder object created inside createConnection? Can we re-use it here?

Reply:
No we cant. This builder is specific to the subscription instead of the Streaming connection.